xmpp-vala/src/module/sasl.vala

```diff
@@ -162,60 +74,34 @@ namespace Xmpp.Sasl {
-                var split = name.split("@");
-                if (split[1] == stream.remote_name.to_string()) {
-                    name = split[0];
-                } else {
-                    use_full_name = true;
-                }
+
+            string? suggested_mechanism = this.sasl_context.client_suggest_mechanism(string.joinv(" ", supported_mechanisms));
+            if (suggested_mechanism == null) {
+                stderr.printf("No supported mechanism provided by server at %s. Offered: %s\n", stream.remote_name.to_string(), string.joinv(",", supported_mechanisms));
+                return;
             }
-            string name = this.name;
-            if (!use_full_name && name.contains("@")) {
-                var split = name.split("@");
-                if (split[1] == stream.remote_name.to_string()) {
-                    name = split[0];
-                }
+
+            var flag = new Flag();
+            Gsasl.Result result;
+            result = this.sasl_context.client_start(suggested_mechanism, out flag.sasl_session);
+            if (result != Gsasl.Result.OK) {
+                sasl_error(stream, result.description());
+                return;
             }
-            if (Mechanism.SCRAM_SHA_1 in supported_mechanisms) {
-                string normalized_password = password.normalize(-1, NormalizeMode.NFKC);
-                string client_nonce = Random.next_int().to_string("%.8x") + Random.next_int().to_string("%.8x") + Random.next_int().to_string("%.8x");
-                string initial_message = @"n=$name,r=$client_nonce";
-                stream.write(new StanzaNode.build("auth", NS_URI).add_self_xmlns()
-                        .put_attribute("mechanism", Mechanism.SCRAM_SHA_1)
-                        .put_node(new StanzaNode.text(Base64.encode((uchar[]) ("n,,"+initial_message).to_utf8()))));
-                var flag = new Flag();
-                flag.mechanism = Mechanism.SCRAM_SHA_1;
-                flag.name = name;
-                flag.password = normalized_password;
-                flag.client_nonce = client_nonce;
-                stream.add_flag(flag);
-            } else if (Mechanism.PLAIN in supported_mechanisms) {
-                stream.write(new StanzaNode.build("auth", NS_URI).add_self_xmlns()
-                                    .put_attribute("mechanism", Mechanism.PLAIN)
-                                    .put_node(new StanzaNode.text(Base64.encode(get_plain_bytes(name, password)))));
-                var flag = new Flag();
-                flag.mechanism = Mechanism.PLAIN;
-                flag.name = name;
-                stream.add_flag(flag);
-            } else {
-                stderr.printf("No supported mechanism provided by server at %s\n", stream.remote_name.to_string());
+            flag.sasl_session.set_property(Gsasl.Property.AUTHID, name);
+            flag.sasl_session.set_property(Gsasl.Property.HOSTNAME, stream.remote_hostname);
```

Comment:
According to XEP 233, the hostname for GSSAPI is provided with the SASL mechanisms, see https://xmpp.org/extensions/xep-0233.html#examples
Can we use this instead of passing the hostname around everywhere?

Reply:
I've updated this to use the hostname from the SASL mechanism, and reverted the passing around of hostnames everywhere.